File: lib/core/errors/exceptions.dart
class AppException implements Exception {
  final String code;
  final String message;
  final List<ErrorDetail> details;

  AppException({
    required this.code,
    required this.message,
    this.details = const [],
  });

  String get userMessage {
    switch (code) {
      case 'AUTH_ERROR':
      case 'TOKEN_EXPIRED':
        return 'Sesi kamu sudah habis. Silakan login kembali.';
      case 'INVALID_CREDENTIALS':
        return 'Email atau password yang kamu masukkan salah. Periksa kembali data kamu.';
      case 'FORBIDDEN':
        return 'Kamu tidak punya akses untuk melakukan ini.';
      case 'VALIDATION_ERROR':
        if (details.isNotEmpty) {
          return details.map((d) => d.message).join('\n');
        }
        return 'Data yang dimasukkan tidak valid. Periksa kembali form kamu.';
      case 'CONFLICT':
        return message.isNotEmpty ? message : 'Data yang kamu masukkan sudah ada.';
      case 'NOT_FOUND':
        return 'Data tidak ditemukan.';
      case 'RATE_LIMITED':
        return 'Terlalu banyak percobaan. Tunggu sebentar lalu coba lagi.';
      case 'NETWORK_ERROR':
        return 'Tidak bisa terhubung ke server. Periksa koneksi internet kamu.';
      case 'TIMEOUT':
        return 'Koneksi timeout. Coba lagi dalam beberapa saat.';
      default:
        return message.isNotEmpty ? message : 'Terjadi kesalahan. Coba lagi.';
    }
  }

  @override
  String toString() => 'AppException($code: $message)';
}

class ErrorDetail {
  final String field;
  final String message;

  ErrorDetail({required this.field, required this.message});
}

File: backend/app/core/exceptions.py
from fastapi import HTTPException


class AppException(HTTPException):
    def __init__(self, code: str, message: str, status_code: int = 500, details: list = None):
        self.code = code
        self.message = message
        self.details = details or []
        super().__init__(status_code=status_code, detail=message)


class AuthenticationError(AppException):
    def __init__(self, message: str = "Autentikasi diperlukan"):
        super().__init__(code="AUTH_ERROR", message=message, status_code=401)


class InvalidCredentialsError(AppException):
    def __init__(self):
        super().__init__(code="INVALID_CREDENTIALS", message="Email atau password salah", status_code=401)


class ForbiddenError(AppException):
    def __init__(self, message: str = "Kamu tidak punya akses"):
        super().__init__(code="FORBIDDEN", message=message, status_code=403)


class NotFoundError(AppException):
    def __init__(self, resource: str = "Data", resource_id: str = ""):
        msg = f"{resource} tidak ditemukan" + (f": {resource_id}" if resource_id else "")
        super().__init__(code="NOT_FOUND", message=msg, status_code=404)


class ConflictError(AppException):
    def __init__(self, message: str = "Data sudah ada"):
        super().__init__(code="CONFLICT", message=message, status_code=409)


class ValidationError(AppException):
    def __init__(self, details: list):
        super().__init__(code="VALIDATION_ERROR", message="Data yang dimasukkan tidak valid", status_code=422, details=details)


class RateLimitError(AppException):
    def __init__(self, retry_after: int = 60):
        super().__init__(code="RATE_LIMITED", message=f"Terlalu banyak percobaan. Coba lagi dalam {retry_after} detik.", status_code=429)
